src/nspawn/nspawn.c

```diff
@@ -6154,7 +6155,16 @@ static int run(int argc, char *argv[]) {
                                 goto finish;
                         }
 
-                        if (access_nofollow(p, F_OK) < 0) {
+                        check_usr = getenv_bool("SYSTEMD_NSPAWN_CHECK_USR");
+                        if (check_usr < 0 && check_usr != -ENXIO) {
+                                r = log_error_errno(check_usr, "Failed to parse $SYSTEMD_NSPAWN_CHECK_USR: %m");
+                                goto finish;
+                        }
+
+                        is_usr_present = access_nofollow(p, F_OK) >= 0;
+                        if (is_usr_present == 0 && check_usr == 0)
+                                log_debug("Directory %s is missing a /usr directory, continuing anyway.", p);
+                        else if (is_usr_present == 0) {
                                 r = log_error_errno(SYNTHETIC_ERRNO(EINVAL),
                                                     "Directory %s doesn't look like it has an OS tree (/usr/ directory is missing). Refusing.", arg_directory);
                                 goto finish;
```

Comment:
hmm, any chance you can split this whole check (i.e. starting with the `path_join()` calls above into a new helper placed somewhere further up called `validate_usr_directory()` or something like that?

so that we only call here:

```c
r = validate_usr_directory(arg_directory, arg_pivot_root_new);
if (r < 0)
        goto finish;
```

After all main() is really long enough already, and if this check now grows like it does it really better be moved out of it.